crates/matrix-sdk-base/src/response_processors/room/sync_v2.rs

```diff
@@ -61,39 +62,48 @@ pub async fn update_joined_room(
     room_info.mark_state_fully_synced();
     room_info.handle_encryption_state(requested_required_states.for_room(room_id));
 
-    let (raw_state_events, state_events) = state_events::sync::collect(&joined_room.state.events);
-
     let mut new_user_ids = BTreeSet::new();
 
-    state_events::sync::dispatch(
-        context,
-        (&raw_state_events, &state_events),
-        &mut room_info,
-        ambiguity_cache,
-        &mut new_user_ids,
-        state_store,
-    )
-    .await?;
+    let state = match joined_room.state {
+        State::Before(state) => {
+            let (raw_state_events, state_events) = state_events::sync::collect(&state.events);
+            state_events::sync::dispatch(
+                context,
+                (&raw_state_events, &state_events),
+                &mut room_info,
+                ambiguity_cache,
+                &mut new_user_ids,
+                state_store,
+            )
+            .await?;
+
+            let (raw_state_events_from_timeline, state_events_from_timeline) =
+                state_events::sync::collect_from_timeline(&joined_room.timeline.events);
+            state_events::sync::dispatch(
+                context,
+                (&raw_state_events_from_timeline, &state_events_from_timeline),
+                &mut room_info,
+                ambiguity_cache,
+                &mut new_user_ids,
+                state_store,
+            )
+            .await?;
```

Comment:
Out of curiosity: could this part be *after* the `match`, if each match arm returned `raw_state_events` + `state_events`?

Reply:
No, because the other variant is `State::After` (I will add support for that in another PR), where we **don't** process the state events in the timeline.

Reply:
The behavior to apply is detailed in the docs of the variant, fwiw.